### PR TITLE
Add JS Bundling

### PR DESCRIPTION
**Pull request by tisba**
This PR adds the ability to bundle a test case using ECMAScript modules (`import`/`export`). The goal is to enable an effortless modularisation of test cases, without the need for an external tool, like webpack or gulp. This change will make https://github.com/stormforger/gulp-template obsolete.

## Changes

The **new sub command** `forge test-case build <input>` can be used to bundle a test case. This is intended to be used for inspecting the output or committing it into a SCM.

All other commands that handle updating or creating test cases, will automatically try to bundle based on the given input, **if the entry file has the file extension `.mjs`**. Files being included do not need this extension though.

### Parameterising at Build Time

Using replacements, parameterisation at build time becomes possible:

```terminal
forge test-case build testdata/cases/index.js --define='replacements.ENV="production"' --define="replacements.debug=true"
```

Given

```javascript
const config = {
  env: replacements.ENV,
  debugMode: replacements.debug,
}
```

will be replaced to:

```javascript
const config = {
  env: "production",
  debugMode: true,
}
```

Note that replacements are literal, so the double quotes around `production` are required.

## Steps/TODOs

- [x] Source Map Support
- [x] ~~there should be an opt-out option/flag, like `--disable-bundling`~~ done by only touching `.mjs` entry points
- [x] ~~use test corpus of existing, valid cases and check for any semantic differences (sanity check)~~
- [x] decide how to name the replacement parameter (currently `--define`)
- Add bundling to endpoints
  - [x] `test-case validate`
  - [x] `test-case update`
  - [x] `test-case create`
  - [x] `test-case launch`
- [x] create simple example
- [x] replacement -> define
- [x] check implementation
- [x] check docs
- [x] add READMEs in `examples/`
- [x] Do all `forge tc` subcommands support the `--define` flag, where applicable

### CODE REVIEW

**Review by zeisss — CHANGES_REQUESTED** [review]

File: cmd/build.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+	buildOpts struct {
+		Replacements []string
+	}
+)
+
+func init() {
+	RootCmd.AddCommand(buildCmd)
+
+	buildCmd.PersistentFlags().StringArrayVar(&buildOpts.Replacements, "define", []string{}, "Substitute a list of K=V while parsing: debug=false")
```
Comment (zeisss) [inline cmd/build.go:30]: Take a look at https://godoc.org/github.com/spf13/pflag#StringToStringVar

Comment (tisba) [reply cmd/build.go:28]: I did not got a value passed in containing quotes properly.

Using

```golang
buildCmd.PersistentFlags().StringToStringVar(&buildOpts.Replacements, "define", nil, "Substitute a list of K=V while parsing: debug=false")
```

With `--define='replacements.ENV="production"'` seems to drop the last double quotes. I've also tried different variations of escaping.

Any idea?

Comment (zeisss) [reply cmd/build.go:28]: I think that is on purpose as it the way to quote the value. If we want to keep the quote, we maybe go back to do it manually. 😞

**Review by zeisss — DISMISSED** [review]

File: testdata/cases/modules/options.js [hunk 2]
```diff
@@ -0,0 +1,9 @@
+definition.addTarget("testapp.loadtest.party")
+
+definition.setArrivalPhases([{
+    duration: 60,
+    rate: 42,
+    max_clients: 23,
+  },
+]);
+definition.foo(); // Invalid
```
[the remark this answers is not in the document]
Comment (tisba) [reply testdata/cases/modules/options.js:9]: We can have an "invalid" module for testing maybe.

File: cmd/testcase_validate.go [hunk 3]
```diff
@@ -110,11 +110,16 @@ func runTestCaseValidateArg(cmd *cobra.Command, client *api.Client, fileOrStdin
 		return !success, nil
 	}
 
-	errorMeta, err := api.UnmarshalErrorMeta(strings.NewReader(message))
+	errorMeta, err := api.ErrorDecoder{SourceMapper: mapper}.UnmarshalErrorMeta(strings.NewReader(message))
 	if err != nil {
 		return true, err
 	}
 
+	if len(errorMeta.Errors) > 0 && errorMeta.Errors[0].EvaluationErrorMeta != nil {
```
Comment (zeisss) [inline cmd/testcase_validate.go:118]: @tisba This can probably be deleted, right? Feels like this was just for debugging?

Comment (tisba) [reply cmd/testcase_validate.go:118]: yes, just to poke around 👍

File: cmd/testcase_update_test.go [hunk 4]
```diff
@@ -24,7 +24,7 @@ func TestPrintValidationResultHuman_ValidationErrors(t *testing.T) {
 			{Code: "E0", Title: "Validation Error"},
 		},
 	})
-	assert.Equal(t, buf.String(), "# FILE: test.js\nWARN: TestCase updated, but validation errors occured.\n\n1) E0: Validation Error\n\n")
+	assert.Equal(t, buf.String(), "WARN: TestCase updated, but validation errors occured.\n\n1) E0: Validation Error\n")
```
Comment (tisba) [inline cmd/testcase_update_test.go:27]: For non-eval errors and for `test-case validate` we could run into problems if we validate multiple test cases via `forge test-case validate *.js`. In this case we need to know which file did contain the issues.

File: cmd/testcase_create.go [hunk 5]
```diff
@@ -79,7 +79,7 @@ func init() {
 func runTestCaseCreate(cmd *cobra.Command, args []string) {
 	orgaUID := testCaseCreateOpts.Organisation
 
-	fileName, testCaseFile, err := readTestCaseFromStdinOrReadFromArgument(args[1], "test_case.js")
+	result, err := testCaseFileBundler{}.bundle(args[1], "test_case.js")
```
Comment (tisba) [inline cmd/testcase_create.go:82]: `result` -> `bundle`

File: api/meta.go [hunk 6]
```diff
@@ -68,27 +71,35 @@ func UnmarshalMeta(input io.Reader) (Meta, error) {
...
 				log.Fatal(err)
 			}
 
-			data.Errors[i].FormattedError = errorMeta.(*EvaluationErrorMeta).String()
+			if dec.SourceMapper != nil {
+				errorMeta.Stack = updateStackFramesFromSourceMapper(errorMeta.Stack, dec.SourceMapper)
+			}
+
+			data.Errors[i].FormattedError = errorMeta.String()
+			data.Errors[i].EvaluationErrorMeta = errorMeta
```
Comment (tisba) [inline api/meta.go:102]: We can drop this.

File: cmd/testcase_validate.go (hunk 3 above)
Comment (tisba) [reply cmd/testcase_validate.go:119]: 🔥

File: cmd/build.go [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+	defines := make(map[string]string)
+	for _, kv := range buildOpts.Replacements {
+		equals := strings.IndexByte(kv, '=')
+		if equals == -1 {
+			log.Fatalf("Missing \"=\": %q", kv)
+		}
+
+		defines[kv[:equals]] = kv[equals+1:]
+	}
```
Comment (tisba) [inline cmd/build.go:82]: Move to custom cobra argument type?

Comment (zeisss) [reply cmd/build.go]: done

File: cmd/testcase_update_test.go (hunk 4 above)
Comment (zeisss) [reply cmd/testcase_update_test.go:27]: I opted to add the filename output only to the `forge test-case validate`. I don't think the filename adds much information to the other commands.

**Comment by zeisss** [comment]
@tisba wanna give it a try? Generally I think everything works.

**Comment by tisba** [comment]
```
RequestError: Path must be a non empty String!
    at new RequestError (custom_validation_errors.js:48:17)
    at Context.request (test_case_dsl.js:489:11)
    at Context.get (test_case_dsl.js:463:10)
    at sandbox/complex/components/cart.js:7:12
    at sandbox/complex/epp/crawler.js:7:6
    at Definition.session (test_case_dsl.js:97:3)
    at sandbox/complex/epp.mjs:17:11
```

🤔 

I think we should mark the internal frames. Maybe by making it clear in the name? We could even hide them (by default).

**Comment by zeisss** [comment]
> I think we should mark the internal frames. 

At least from your example it was pretty clear, but yeah.  Not sure about hiding them - they help with showing context what gets called.

I kind of miss the function names from the non-internal frames though - is this a bug?

**Comment by tisba** [comment]
> > I think we should mark the internal frames.
> 
> At least from your example it was pretty clear, but yeah. Not sure about hiding them - they help with showing context what gets called.
> 
> I kind of miss the function names from the non-internal frames though - is this a bug?

The internal frames are only useful for us for debugging (could be opt-in with a flag e.g.).

I got the function names back, by only using the mapped version if the name/context is not empty. Still checking if this is intended/expected though...

File: cmd/testcase_update_test.go (hunk 4 above)
Comment (tisba) [reply cmd/testcase_update_test.go:27]: Yeah. In all other cases it can only reference one "file".

File: cmd/testcase_validate.go [hunk 8]
```diff
@@ -36,6 +38,10 @@ Examples
 
 	forge test-case validate acme-inc ./dist/foo.js ./dist/bar.js ./dist/foobar.js
 
+Bundling
+--------
+
+Validate automatically bundles your javascript file, if you use the .mjs extension. See 'forge build' for more details.
```
Comment (tisba) [inline cmd/testcase_validate.go:44]: Does it make sense to extract this usage hints into a variable? So that it stays consistent everywhere?

Comment (zeisss) [reply cmd/testcase_validate.go:44]: Thats a good idea 👍

Comment (zeisss) [reply cmd/testcase_validate.go:44]: But the first word is different for each command ..

Comment (tisba) [reply cmd/testcase_validate.go:44]: hmm, true. Maybe we can rephrase that a bit. Let's see.

**Comment by tisba** [comment]
How do we feel about the name of the "replacements" flag, @zeisss? `--define` sounds okay, but is not perfectly on point. `--replacement` feels more accurate, but is rather long. Any other ideas?

**Comment by zeisss** [comment]
> How do we feel about the name of the "replacements" flag, @zeisss?

I am ok with it. I explicitly mentioned esbuild in the docstring and the feature is named the same there. Might make it easier for understanding.

Also, you define variables, you CANNOT replace any defined variable. thus replacement might be misleading.

File: cmd/testcase_build.go [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+'forge test-case build' will compile a single JavaScript out of it, resolving the
+imports transparently and adding defined variables, if used.
+
+This is also done automatically for you when using other 'forge test-case'
+commands, so you don't need to call this command directly.
+
+Imports (ECMAScript modules)
+----------------------------
+Using 'forge test-case build' allows importing other JavaScript files via the 'import'
+statement, if your first files ends in '.mjs':
```
Comment (grekko) [inline cmd/testcase_build.go:38]:
```suggestion
statement, if the entry file ends in '.mjs':
```

**Pull request merged**